legion/engine/physics/halfedgeface.cpp

```diff
@@ -0,0 +1,178 @@
+        };
+
+        forEachEdge(initializeEdgeToFaceFunc);
+
+        faceCount++;
+    }
+
+    float HalfEdgeFace::calculateFaceExtents()
+    {
+        //get vector of vertices of face
+        std::vector<math::vec3> vertices;
+        vertices.reserve(6);
+
+        auto collectVertices = [&vertices](HalfEdgeEdge* edge) {vertices.push_back(edge->edgePosition); };
+        forEachEdge(collectVertices);
+
+        //get tangents of normals
+        math::vec3 forward = math::normalize(centroid - startEdge->edgePosition);
+        math::vec3 right = math::cross(normal, forward);
+
+        math::vec3 maxForward, minForward, maxRight, minRight;
+
+        //get support point for the tangents and inverse tangents of this face 
+        PhysicsStatics::GetSupportPoint(vertices, forward, maxForward);
+        PhysicsStatics::GetSupportPoint(vertices, -forward, minForward);
+
+        PhysicsStatics::GetSupportPoint(vertices, right, maxRight);
+        PhysicsStatics::GetSupportPoint(vertices, -right, minRight);
+
+        float maxForwardLength = math::dot(maxForward - centroid, forward);
+        float minForwardLength = math::dot(minForward - centroid, -forward);
+
+        float maxRightLength = math::dot(maxRight - centroid, right);
+        float minRightLength = math::dot(minRight - centroid, -right);
+
+        return (maxForwardLength + minForwardLength) + (maxRightLength + minRightLength);
+    }
+
+    void HalfEdgeFace::forEachEdge(legion::core::delegate< void(HalfEdgeEdge*)> functionToExecute,
+        legion::core::delegate <HalfEdgeEdge* (HalfEdgeEdge*)> getNextEdge )
+    {
+        HalfEdgeEdge* initialEdge = startEdge;
+        HalfEdgeEdge* currentEdge = startEdge;
+
+        if (!currentEdge) return;
+
+        //the HalfEdgeEdge* 'startEdge' creates a ring buffer.
+        //This means that initialEdge will eventually go back to "startEdge", ending the loop.
+        do
+        {
+            HalfEdgeEdge* edgeToExecuteOn = currentEdge;
+            currentEdge = getNextEdge(currentEdge);
+            functionToExecute(edgeToExecuteOn);
+
+        } while (initialEdge != currentEdge && getNextEdge(currentEdge) != nullptr);
+    }
+
+    void HalfEdgeFace::forEachEdgeReverse(legion::core::delegate<void(HalfEdgeEdge*)> functionToExecute)
+    {
+        auto getPrevEdges = [](HalfEdgeEdge* current) {return current->prevEdge; };
+
+        forEachEdge(functionToExecute, getPrevEdges);
+    }
+
+    void HalfEdgeFace::inverse()
+    {
+        normal = -normal; // Inverse the normal
+
+        //collect edges into std::vector
+        std::vector<HalfEdgeEdge*> edges;
+
+        auto collectEdges = [&edges](HalfEdgeEdge* edge)
+        {
+            edges.push_back(edge);
+        };
+
+        forEachEdgeReverse(collectEdges);
+
+        for (size_t i = 0; i < edges.size(); i++)
+        {
+            int nextIndex =( i + 1) % edges.size();
+            int prevIndex = (i - 1) == -1 ? edges.size()-1 : (i - 1);
```

Comment:
I'll convert it i to int since it would be much more readable

Reply:
actually that was the wrong thing to take away here, since `.size()` will always return a `size_type` anyways. turning i into an int will just cause signed unsigned comparison, and implicit casts from `size_type` to `int`. actually `nextIndex` and `prevIndex` should also be `size_type`.

generally if you're dealing with indices and sizes and amounts, you want to be using `size_type`